File: 2022/day_03.py
from itertools import chain

import aoc_lube
from aoc_lube.utils import chunk

RUCKSACKS = aoc_lube.fetch(year=2022, day=3).splitlines()

def priority(letter):
    offset = ord("a") if letter.islower() else (ord("A") - 26)
    return 1 + ord(letter) - offset

def part_one():
    return sum(map(priority,
        chain.from_iterable(
            set(rucksack[:(l := len(rucksack) // 2)]) & set(rucksack[l:])
            for rucksack in RUCKSACKS
        )
    ))

def part_two():
    return sum(map(priority,
        chain.from_iterable(
            set(a) & set(b) & set(c)
            for a, b, c in chunk(RUCKSACKS, 3)
        )
    ))

aoc_lube.submit(year=2022, day=3, part=1, solution=part_one)
aoc_lube.submit(year=2022, day=3, part=2, solution=part_two)

File: 2022/day_03.nim
include aoc

let sacks = fetch(2022, 3).splitLines()

proc priority(letter: char): int =
  let offset = letter.isLowerAscii ? 'a'.ord: 'A'.ord - 26
  1 + letter.ord - offset

part 1: sum sacks
  .mapIt(
    it
    .toSeq
    .distribute(2)
    .mapIt(it.toHashSet)
    .foldl(a * b)
    .pop
    .priority
  )

part 2: sum sacks
  .distribute(sacks.len div 3)
  .mapIt(
    it
    .mapIt(it.toHashSet)
    .foldl(a * b)
    .pop
    .priority
  )